Backend/src/main/java/com/SYSC4806_Group13/SYSC4806_Project/Controllers/CartItemRestController.java

```diff
@@ -28,73 +29,123 @@ public CartItemRestController(CartItemRepository cartItemRepo, ListingRepository
     @ResponseBody()
-    public List<CartItem> getUserCartItems(@CurrentUser UserPrincipal userPrincipal) {
-        return cartItemRepo.findAllByUserId(userPrincipal.getId());
+    public List<CartItem> getUserCurrentCartItems(@CurrentUser UserPrincipal userPrincipal) {
+        return cartItemRepo.findAllByUserIdAndPurchaseDateTimeIsNull(userPrincipal.getId());
+    }
+
+    //Given user and listing, return quantity and inventory
+    @GetMapping("/cartItem")
+    @ResponseBody()
+    public Map<String, Integer> getCartItem(@CurrentUser UserPrincipal userPrincipal, @RequestParam(value = "listingId") long listingId) {
+        Long userId = (Long) getValidatedAttribute_NonNull("userId", userPrincipal.getId());
+        Long validatedListingId = (Long) getValidatedAttribute_NonNull("listingId", listingId);
+
+        CartItem ci = cartItemRepo.findCartItemByUserIdAndListing_ListingIdAndPurchaseDateTimeIsNull(userId, validatedListingId);
+        Map<String, Integer> map = new HashMap<String, Integer>();
+        if (ci == null) {
+            map.put("quantity", 0);
+            map.put("inventory", listingRepo.findByListingId(validatedListingId).getInventory());
+        } else {
+            map.put("quantity", ci.getQuantity().intValue());
+            map.put("inventory", ci.getListing().getInventory());
+        }
+        return map;
     }
 
     @PostMapping("/cartItems")
-    public Map<String, Object> postUserCartItem(@RequestBody Map<String, Long> payload, @CurrentUser UserPrincipal userPrincipal) {
+    @ResponseBody()
+    public CartItem postUserCartItem(@RequestBody Map<String, Long> payload, @CurrentUser UserPrincipal userPrincipal) {
         Long userId = (Long) getValidatedAttribute_NonNull("userId", userPrincipal.getId());
         Long listingId = (Long) getValidatedAttribute_NonNull("listingId", payload.get("listingId"));
         Long quantity = getValidatedLongAttribute_positiveOnly("quantity", payload.get("quantity"), true);
 
-        if (cartItemRepo.findCartItemByUserIdAndListing_ListingId(userId, listingId) != null) {
+        if (cartItemRepo.findCartItemByUserIdAndListing_ListingIdAndPurchaseDateTimeIsNull(userId, listingId) != null) {
             throw new DuplicateException("A cartItem for this user and listing already exists");
         }
 
         Listing listing = listingRepo.findByListingId(listingId);
         if (listing == null) {
             throw new NotFoundException("ListingId=" + listingId + " does not exist");
         }
+        if (listing.getInventory() < quantity) {
+            throw new BadRequestException("Cart item quantity cannot be greater than current listing inventory of " + listing.getInventory());
+        }
 
         CartItem cartItem = new CartItem(userId, listing, quantity);
         cartItemRepo.save(cartItem);
 
-        Map<String, Object> map = new HashMap<String, Object>();
-        map.put("cartItemId", cartItem.getId());
-
-        return map;
+        return cartItem;
```

Comment:
Does cartItem hold the quantity which is why you return it?

Reply:
Yes, it holds the quantity